Initial commit

beechtom opened the pull request:
This adds a `http_request` task which makes HTTP requests.

Closes puppetlabs/bolt#2103

beechtom committed:
Initial commit

This adds a `http_request` task which makes HTTP requests.

beechtom commented on tasks/init.json:
```diff
@@ -0,0 +1,52 @@
+      "type": "Array[Numeric]",
+      "default": [200]
+    },
+    "url": {
```
Went with `url` over a combination of `base_url` and `path`, as it seems simpler to only have a single parameter that handles the URL and plans can use interpolation to accomplish what `base_url` and `path` would.

adreyer replied:
The purpose of that spec was to support remote targets where the url is configured in inventory and the path is specified for each invocation.

nicklewis replied:
But wouldn't that be under `_target` then?

nicklewis replied:
Ah meaning you would only pass `path` as a task parameter and omit `baseurl`, opting instead to use what was specified in `_target`?

beechtom commented on tasks/init.json:
```diff
@@ -0,0 +1,52 @@
+      "description": "The request body.",
+      "type": "Optional[Data]"
+    },
+    "body_format": {
```
Opted for `body_format` over `json_endpoint`, as it provides a little more flexibility in indicating if the body should be serialized/parsed as JSON while allowing for other common formats as well.

beechtom commented on tasks/init.json:
```diff
@@ -0,0 +1,52 @@
+      "description": "The fully qualified URL scheme to make requests to.",
+      "type": "String[1]"
+    },
+    "user_agent": {
```
Added this since it seems helpful for users who may want to log when Bolt is making requests.

beechtom commented on tasks/init.rb:
```diff
@@ -0,0 +1,137 @@
+  # Formats the body for the request. If the Content-Type is application/json,
+  # this will convert the body to JSON. Otherwise, the body will be stringified.
+  #
+  def make_body(body, body_format)
```
`body_format` will set a default `Content-Type` header, which can be overridden by setting the same header under `headers`. I'm not sure if it's more appropriate to check the `body_format` or `Content-Type` header when determining how to format the body...

adreyer commented on tasks/init.rb:
```diff
@@ -0,0 +1,137 @@
+    json = response['Content-Type'] == 'application/json'
+
+    details = {
+      body:        json ? JSON.parse(response.body) : response.body,
```
A common pain point in dealing with JSON APIs is hitting the wrong route and getting a non-json response. I think it would be better if there was a way to make that an error rather than a completely unexpected structure for the response.

nicklewis commented on tasks/init.rb:
```diff
@@ -0,0 +1,137 @@
+      status_code: response.code
+    }
+
+    if !success_codes.include?(response.code.to_i)
```
The task should always return the response as long as one was returned and leave it up to the caller to decide which responses are acceptable.

nicklewis reviewed (changes requested):
I think this has a bit too much magic when what we really need _first_ is a low-level implementation of the HTTP protocol on which we can build abstractions / conveniences.

For now it might be best to leave building the body and setting the content-type, as well as parsing the response, up to the user. We can build an opinionated JSON in+out task or option separately.

nicklewis commented on tasks/init.json:
```diff
@@ -0,0 +1,41 @@
+    },
+    "headers": {
+      "description": "A map of headers to add to the payload.",
+      "type": "Optional[Hash[String, String]"
```
```suggestion
      "type": "Optional[Hash[String, String]]"
```

This causes an extremely obtuse Puppet error, for the record... "Syntax error at end of input" with no context.

beechtom replied:
🤦

nicklewis commented on tasks/init.rb:
```diff
@@ -0,0 +1,61 @@
+    method  = opts[:method].capitalize.to_sym
+    uri     = URI.parse(opts[:base_url]) + (opts[:path] || '')
+    headers = opts[:headers] || {}
+    options = opts.slice(:cacert, :cert, :key)
```
`slice` doesn't work on Ruby < 2.5. Since this can run on a remote target rather than with Bolt's Ruby, we should aim for better compatibility.

I also ran into an issue with `compact` in the `ruby_task_helper`. That method doesn't work on Ruby < 2.4.

nicklewis reviewed (changes requested):
There's some wonkiness around content encoding. For instance, hitting `https://www.google.com` returns a response with content-type indicating the encoding is `ISO-8859-1` but the Ruby string is binary. Forcing the encoding to ISO-8859-1 and then encoding it to UTF-8 works, but otherwise Bolt will complain.

On the other hand, `https://www.yahoo.com` doesn't specify an encoding but the Ruby string is UTF-8.

We may also want to have an option to follow redirects, as it's a hassle when you just get back a 301 Moved.

beechtom commented:
Updated. The task will now enforce encoding for the body based on the `Content-Type` header (if present) before encoding as UTF-8. Also added an option for following redirects (default to `true`) and an option for setting the max number of redirects (default to `20`, which seems to be a common maximum used in browsers).

I'll look at getting the task helper updated to not use `compact`.

lucywyman commented on CHANGELOG.md:
```diff
@@ -0,0 +1,3 @@
+## Release 1.0.0
```
Probably want this to be `0.1.0` for the initial release https://semver.org/#how-should-i-deal-with-revisions-in-the-0yz-initial-development-phase

beechtom replied:
I have the initial release at `1.0.0` since I'm assuming once this is in it's immediately going to be shipped with Bolt and be ready for use in production: https://semver.org/#how-do-i-know-when-to-release-100 . If we don't immediately ship this and are still considering it 'in development', then I'm fine going down to a `0.x` release.

lucywyman replied:
Right but _we_ aren't currently using it in production, and it seems better to make as few guarantees as possible.

beechtom replied:
I don't really understand what you mean by we aren't using it in production...
Regardless, updated to 0.1.0.

lucywyman commented on Gemfile:
```diff
@@ -0,0 +1,26 @@
+
+group :development do
+  gem "puppet-module-posix-default-r#{minor_version}", require: false, platforms: [:ruby]
+  gem "puppet-module-posix-dev-r#{minor_version}",     require: false, platforms: [:ruby]
```
Add pdk

lucywyman commented on metadata.json:
```diff
@@ -0,0 +1,16 @@
+{
+  "name": "puppetlabs-http_request",
+  "version": "1.0.0",
```
Same update, 0.1.0

lucywyman commented on init.json:
```diff
@@ -0,0 +1,15 @@
+{
```
This seems to be here by accident?

beechtom replied:
Oops, had a duplicate clone of the repo and accidentally pushed from it. Should be better now. @lucywyman

lucywyman commented on tasks/init.json:
```diff
@@ -0,0 +1,51 @@
+{
+  "description": "Make a HTTP or HTTPS request.",
```
It'd be good to specify input_method stdin here, if only to exemplify best practices (and also because we don't use env vars)

lucywyman commented on tasks/init.rb:
```diff
@@ -0,0 +1,101 @@
+require_relative '../../ruby_task_helper/files/task_helper'
+
+require 'net/http'
+require 'openssl'
```
If we expect this to be used in plans and not directly - or really even if we expect it to be used directly - I think it makes sense to wrap this in a plan that ensures these gems are available.

lucywyman commented:
@adreyer or @nicklewis Does either of you want to re-review this before merge and release?
